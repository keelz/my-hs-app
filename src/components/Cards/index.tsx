import * as React from 'react';
import Card from '../Card';
import './Cards.css';
import {
  Card as CardType,
  CardResolution,
  CardExt,
  CardLocale
} from '../../redux/reducers/Cards';

interface CardsProps {}

export interface CardsStateProps {
  activeCardClassName: string;
  cards: CardType[];
}

type Props = CardsProps & CardsStateProps;

const Cards: React.SFC<Props> = props =>
  <div className="Cards">
    <div className="Cards-title">
      {props.activeCardClassName}: {props.cards.length}
    </div>
    <div className="Cards-collection">
      { props.cards.map(card =>
        <Card
          key={card.id}
          ext={CardExt.PNG}
          id={card.id}
          locale={CardLocale.EN}
          resolution={CardResolution.SMALL} />)
      }
    </div>
  </div>;

export default Cards;
