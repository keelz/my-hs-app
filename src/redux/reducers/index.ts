import { combineReducers } from 'redux';
import { responsiveStateReducer as Browser } from 'redux-responsive';
import appReducer from './App';

const rootReducer = combineReducers({
  app: appReducer,
  browser: Browser,
});

export default rootReducer;
