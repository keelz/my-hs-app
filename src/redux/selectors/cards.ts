import { createSelector } from 'reselect';
import HSJSON from '../../common/constants/hsJson';
import { IRootState } from '../Types';
import { ICard } from '../../common/models/Card';
import { composeCardClassNames } from '../../common/utils/card';

export class Accessors {
  /**
   * get active card class name.
   * @param state {IRootState}
   */
  public static getActiveCardClassName(state: IRootState): string {
    return state.Cards.activeClassName;
  }

  /**
   * get all cards.
   * @param state {IRootState}
   */
  public static getCards(state: IRootState): ICard[] {
    return state.Cards.data;
  }

  /**
   * get all Cards filters.
   * @param state {IRootState}
   */
  public static getFilters(state: IRootState): { [key: string]: string } {
    return state.Cards.filters;
  }
}

/**
 * get active card class name.
 * @param state {IRootState}
 */
export const selectActiveCardClassName = (state: IRootState): string =>
  createSelector(
    [Accessors.getActiveCardClassName],
    (className) => className
  )(state);

/**
 * get all unique card class names.
 * @param state {IRootState}
 */
export const selectCardClassNames = (state: IRootState) =>
  createSelector(
    [Accessors.getCards],
    (cards) => composeCardClassNames(cards)
  )(state);

/**
 * get all cards.
 * @param state {IRootState}
 */
export const selectCards = (state: IRootState): ICard[] =>
  createSelector(
    [Accessors.getCards],
    (cards) => cards
  )(state);

/**
 * get cards for active class name.
 * @param state {IRootState}
 */
export const selectCardsWithFilters = (state: IRootState): ICard[] =>
  createSelector(
    [
      Accessors.getActiveCardClassName,
      Accessors.getCards,
      Accessors.getFilters,
    ],
    (
      activeClassName: string,
      cards: ICard[],
      filters: { [key: string]: string }
    ) => cards
      // filter by active class name
      .filter(card => card.cardClass === activeClassName)
      // filter by card cost
      .filter(card => {
        const filter = filters[HSJSON.RESPONSE_PARAMS.COST];
        const response = !!filter
          ? parseInt(filter, 10) === 7
            ? card.cost >= 7
            : card.cost === parseInt(filter, 10)
          : true;
        return response;
      })
      // sort by cost
      .sort((a, b) => a.cost - b.cost)
      // sort by name
      .sort((a, b) => a.cost === b.cost
        ? a.name < b.name
          ? -1
          : 0
        : 0)
  )(state);
