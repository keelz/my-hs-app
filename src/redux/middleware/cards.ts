import { Dispatch } from 'redux';
import { AppMiddlewareApi, IActionType } from '../Types';
import cardsService, { ICardsService } from '../../common/services/cards.service';
import {
  CARDS_SET_ACTIVE_CLASSNAME,
  CARDS_SET_CARDS,
  CARDS_SET_FILTER,
  CARDS_SET_PAGINATION,
} from '../reducers/Cards';

export const composeMiddleware = (service: ICardsService) =>
  (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => {
    switch (action.type) {
      case CARDS_SET_ACTIVE_CLASSNAME:
        return service.handleSetActiveClassname(api)(next)(action);
      case CARDS_SET_CARDS:
        next(action);
        service.resetPagination(api)(action);
        return;
      case CARDS_SET_FILTER:
        return service.handleSetFilter(api)(next)(action);
      case CARDS_SET_PAGINATION:
        return service.handleSetPagination(api)(next)(action);
      default:
        return next(action);
    }
  };

const middleware = composeMiddleware(cardsService);

export default middleware;
