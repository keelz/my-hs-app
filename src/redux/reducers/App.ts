import { IActionType } from '../Types';
import { Dispatch } from 'redux';

/** MODELS */
export type AppState = {
  loaded: boolean;
};

/** FACTORIES */
export const appStateFactory = () => ({
  loaded: false,
});

/** ACTIONS */
export const APP_DID_LOAD = 'APP_DID_LOAD';
export const APP_SET_LOADED = 'APP_SET_LOADED';

/** ACTION CREATORS */
export const appDidLoadAction = () =>
  (dispatch: Dispatch) =>
  dispatch({ type: APP_DID_LOAD });

export const appSetLoadedAction = (loaded: boolean) =>
  (dispatch: Dispatch) =>
  dispatch({
    type: APP_SET_LOADED,
    payload: { loaded },
  });

/** REDUCER */
export default (state: AppState = appStateFactory(), action: IActionType<any>): AppState => {
  const { type } = action;
  switch (type) {
    case APP_SET_LOADED: return {
      loaded: action.payload.loaded,
    };
    default: return state;
  }
};
