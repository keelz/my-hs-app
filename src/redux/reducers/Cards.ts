import { IActionType } from '../Types';
import { Dispatch } from 'redux';
import { IPagination } from '../../common/models/app.model';
import {
  ICard,
  CardClassName
} from '../../common/models/cards.model';

/** STATE MODEL */
export interface ICardsState {
  activeClassName: string;
  data: ICard[];
  filters: { [field: string]: string };
  pagination: IPagination;
}

/** STATE FACTORY */
export const defaultState = (): ICardsState => ({
  activeClassName: CardClassName.DRUID,
  data: Array(0),
  filters: {},
  pagination: {
    currentPage: 0,
    itemsPerPage: 0,
    pages: 0,
    total: 0,
  },
});

/** ACTIONS */
export const CARDS_DELETE_FILTERS = 'CARDS_DELETE_FILTER';
export const CARDS_SET_ACTIVE_CLASSNAME = 'CARDS_SET_ACTIVE_CLASSNAME';
export const CARDS_SET_CARDS = 'CARDS_SET_CARDS';
export const CARDS_SET_FILTER = 'CARDS_SET_FILTER';
export const CARDS_SET_PAGINATION = 'CARDS_SET_PAGINATION';

/** ACTION CREATORS */
export const cardsDeleteFilterAction = (key: string) =>
  (dispatch: Dispatch<any>) =>
  dispatch({
    type: CARDS_DELETE_FILTERS,
    payload: { filters: { [key]: undefined } },
  });

export const cardsSetCardsAction = (data: ICard[]) =>
 (dispatch: Dispatch<IActionType<any>>) =>
 dispatch({
   type: CARDS_SET_CARDS,
   payload: { data },
 });

export const cardsSetActiveClassNameAction = (activeClassName: string) =>
 (dispatch: Dispatch<any>) =>
 dispatch({
   type: CARDS_SET_ACTIVE_CLASSNAME,
   payload: { activeClassName },
 });

export const cardsSetFilterAction = (key: string, value: string) =>
 (dispatch: Dispatch<any>) =>
 dispatch({
   type: CARDS_SET_FILTER,
   payload: { filters: { [key]: value } },
 });

export const cardsSetPaginationAction = (pagination: IPagination) =>
 (dispatch: Dispatch<any>) =>
 dispatch({
   type: CARDS_SET_PAGINATION,
   payload: { pagination },
 });

/** REDUCER */
const reducer = (
  state: ICardsState = defaultState(),
  action: IActionType<ICardsState>): ICardsState => {
  const { type } = action;
  switch (type) {
    case CARDS_DELETE_FILTERS:
      const deleteFilters = { ...state.filters };
      !!action.payload && Object
        .keys(action.payload.filters)
        .forEach(key => delete deleteFilters[key]);
      return {
        ...state,
        filters: { ...deleteFilters },
      };
    case CARDS_SET_ACTIVE_CLASSNAME: return {
      ...state,
      activeClassName: !!action.payload
        ? action.payload.activeClassName
        : CardClassName.DRUID,
    };
    case CARDS_SET_CARDS: return {
      ...state,
      data: !!action.payload
        ? action.payload.data
        : Array(0),
    };
    case CARDS_SET_FILTER:
      const filters = action.payload
        ? { ...state.filters, ...action.payload.filters }
        : { ...state.filters };
      return { ...state, filters };
    case CARDS_SET_PAGINATION: return {
      ...state,
      pagination: !!action.payload
        ? action.payload.pagination
        : state.pagination,
    };
    default: return state;
  }
};

export default reducer;
