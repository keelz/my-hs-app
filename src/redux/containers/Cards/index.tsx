import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import Cards from '../../../components/Cards';
import { RootState } from '../..//Types';
import { Card } from '../../reducers/Cards';

const getCards = (state: RootState) => state.Cards.data;

const getActiveClassName = (state: RootState) => state.Cards.activeClassName;

const cardsSelector = createSelector(
  [getCards, getActiveClassName],
  (cards: Card[], activeClassName) =>
    cards
      .filter(c => c.cardClass === activeClassName)
      .sort((a, b) => a.name < b.name ? - 1 : 0)
);

const mapStateToProps = (state: RootState) => ({
  activeCardClassName: getActiveClassName(state),
  cards: cardsSelector(state),
});

const ConnectedCards = connect(mapStateToProps)(Cards);

export default ConnectedCards;
