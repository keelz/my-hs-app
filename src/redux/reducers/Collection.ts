import { IActionType, ModalState } from '../Types';
import { Dispatch } from 'redux';
import { IPagination } from '../../common/models/App.model';
import {
  ICard,
  CardClassName
} from '../../common/models/Cards.model';

/** STATE MODEL */
export interface ICollectionState {
  activeCard?: ICard;
  activeClassName: string;
  filters: { [field: string]: string };
  modal: ModalState;
  pagination: IPagination;
}

/** STATE FACTORY */
export const defaultState = (): ICollectionState => ({
  activeCard: undefined,
  activeClassName: CardClassName.DRUID,
  filters: {},
  modal: ModalState.CLOSED,
  pagination: {
    currentPage: 0,
    itemsPerPage: 0,
    pages: 0,
    total: 0,
  },
});

/** ACTIONS */
export const COLLECTION_DELETE_FILTERS = 'COLLECTION_DELETE_FILTERS';
export const COLLECTION_SET_ACTIVE_CARD = 'COLLECTION_SET_ACTIVE_CARD';
export const COLLECTION_SET_ACTIVE_CLASSNAME = 'COLLECTION_SET_ACTIVE_CLASSNAME';
export const COLLECTION_SET_FILTER = 'COLLECTION_SET_FILTER';
export const COLLECTION_SET_MODAL = 'COLLECTION_SET_MODAL';
export const COLLECTION_SET_PAGINATION = 'COLLECTION_SET_PAGINATION';

/** ACTION CREATORS */
export const collectionDeleteFilterAction = (key: string) =>
  (dispatch: Dispatch<any>) =>
  dispatch({
    type: COLLECTION_DELETE_FILTERS,
    payload: { filters: { [key]: undefined } },
  });

export const collectionSetActiveCardAction = (activeCard: ICard) =>
  (dispatch: Dispatch<IActionType<any>>) =>
  dispatch({
    type: COLLECTION_SET_ACTIVE_CARD,
    payload: { activeCard },
  });

export const collectionSetActiveClassNameAction = (activeClassName: string) =>
 (dispatch: Dispatch<any>) =>
 dispatch({
   type: COLLECTION_SET_ACTIVE_CLASSNAME,
   payload: { activeClassName },
 });

export const collectionSetFilterAction = (key: string, value: string) =>
 (dispatch: Dispatch<any>) =>
 dispatch({
   type: COLLECTION_SET_FILTER,
   payload: { filters: { [key]: value } },
 });

export const collectionSetModalAction = (modal: ModalState) =>
 (dispatch: Dispatch<IActionType<any>>) =>
 dispatch({
   type: COLLECTION_SET_MODAL,
   payload: { modal },
 });

export const collectionSetPaginationAction = (pagination: IPagination) =>
 (dispatch: Dispatch<any>) =>
 dispatch({
   type: COLLECTION_SET_PAGINATION,
   payload: { pagination },
 });

/** REDUCER */
const reducer = (
  state: ICollectionState = defaultState(),
  action: IActionType<ICollectionState>): ICollectionState => {
  const { type } = action;
  switch (type) {
    case COLLECTION_DELETE_FILTERS:
      const deleteFilters = { ...state.filters };
      !!action.payload && Object
        .keys(action.payload.filters)
        .forEach(key => delete deleteFilters[key]);
      return {
        ...state,
        filters: { ...deleteFilters },
      };
    case COLLECTION_SET_ACTIVE_CARD: return {
      ...state,
      activeCard: action.payload!.activeCard,
    };
    case COLLECTION_SET_ACTIVE_CLASSNAME: return {
      ...state,
      activeClassName: !!action.payload
        ? action.payload.activeClassName
        : CardClassName.DRUID,
    };
    case COLLECTION_SET_FILTER:
      const filters = action.payload
        ? { ...state.filters, ...action.payload.filters }
        : { ...state.filters };
      return { ...state, filters };
    case COLLECTION_SET_MODAL: return {
      ...state,
      modal: action.payload!.modal,
    };
    case COLLECTION_SET_PAGINATION: return {
      ...state,
      pagination: !!action.payload
        ? action.payload.pagination
        : state.pagination,
    };
    default: return state;
  }
};

export default reducer;
