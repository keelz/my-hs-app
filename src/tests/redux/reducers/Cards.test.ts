import * as Reducer from '../../../redux/reducers/Cards';
import mockData from '../../../common/mocks/collection';
import { ICard } from '../../../common/models/cards.model';

describe('Cards Reducer', () => {
  describe('snapshots', () => {
    it('composes default state correctly', () => {
      const { defaultState } = Reducer;
      const result = defaultState();
      expect(result).toMatchSnapshot();
    });

    it('composes state correctly', () => {
      const { defaultState } = Reducer;
      const result = Reducer.default(defaultState(), { type: 'TEST' });
      expect(result).toMatchSnapshot();
    });
  });

  describe('action creators', () => {
    it(`dispatches ${Reducer.CARDS_DELETE_FILTERS} correctly`, () => {
      const dispatch = jest.fn();
      const {
        CARDS_DELETE_FILTERS,
        cardsDeleteFilterAction,
      } = Reducer;
      cardsDeleteFilterAction('test')(dispatch);
      expect(dispatch).toHaveBeenCalledWith({
        type: CARDS_DELETE_FILTERS,
        payload: {
          filters: { test: undefined },
        },
      });
    });

    it(`dispatches ${Reducer.CARDS_SET_CARDS} correctly`, () => {
      const dispatch = jest.fn();
      const {
        cardsSetCardsAction,
        CARDS_SET_CARDS,
      } = Reducer;
      cardsSetCardsAction(mockData as ICard[])(dispatch);
      expect(dispatch).toHaveBeenCalledWith({
        type: CARDS_SET_CARDS,
        payload: { data: mockData },
      });
    });

    it(`dispatches ${Reducer.CARDS_SET_ACTIVE_CLASSNAME} correctly`, () => {
      const dispatch = jest.fn();
      const {
        cardsSetActiveClassNameAction,
        CARDS_SET_ACTIVE_CLASSNAME,
      } = Reducer;
      cardsSetActiveClassNameAction('test')(dispatch);
      expect(dispatch).toHaveBeenCalledWith({
        type: CARDS_SET_ACTIVE_CLASSNAME,
        payload: { activeClassName: 'test' },
      });
    });

    it(`dispatches ${Reducer.CARDS_SET_FILTER} correctly`, () => {
      const dispatch = jest.fn();
      const {
        CARDS_SET_FILTER,
        cardsSetFilterAction,
      } = Reducer;
      cardsSetFilterAction('test', 'test')(dispatch);
      expect(dispatch).toHaveBeenCalledWith({
        type: CARDS_SET_FILTER,
        payload: {
          filters: { test: 'test' },
        },
      });
    });
  });

  describe('reducer', () => {
    it(`handles ${Reducer.CARDS_DELETE_FILTERS} correctly`, () => {
      const {
        defaultState,
        CARDS_DELETE_FILTERS,
      } = Reducer;
      const action = {
        type: CARDS_DELETE_FILTERS,
        payload: {
          ...defaultState(),
          filters: {
            test: 'test',
          },
        },
      };
      const activeState = {
        ...defaultState(),
        filters: { test: 'test', static: 'static' },
      };
      const result = Reducer.default(activeState, action);
      expect(result).toEqual({
        ...defaultState(),
        filters: { static: 'static' },
      });
    });

    it(`handles ${Reducer.CARDS_SET_ACTIVE_CLASSNAME} correctly`, () => {
      const action = {
        type: Reducer.CARDS_SET_ACTIVE_CLASSNAME,
        payload: {
          ...Reducer.defaultState(),
          activeClassName: 'test',
        },
      };
      const result = Reducer.default(undefined, action);
      expect(result).toEqual({
        ...Reducer.defaultState(),
        activeClassName: 'test',
      });
    });

    it(`handles ${Reducer.CARDS_SET_ACTIVE_CLASSNAME} with null payload correctly`, () => {
      const action = { type: Reducer.CARDS_SET_ACTIVE_CLASSNAME };
      const result = Reducer.default(undefined, action);
      expect(result).toEqual({ ...Reducer.defaultState() });
    });

    it(`handles ${Reducer.CARDS_SET_CARDS} correctly`, () => {
      const testData = mockData.slice(0, 3) as ICard[];
      const action = {
        type: Reducer.CARDS_SET_CARDS,
        payload: {
          ...Reducer.defaultState(),
          data: [...testData],
        },
      };
      const result = Reducer.default(undefined, action);
      expect(result).toEqual({
        ...Reducer.defaultState(),
        data: [...testData],
      });
    });

    it(`handles ${Reducer.CARDS_SET_CARDS} with null payload correctly`, () => {
      const action = {
        type: Reducer.CARDS_SET_CARDS,
      };
      const result = Reducer.default(undefined, action);
      expect(result).toEqual({ ...Reducer.defaultState() });
    });

    it(`handles ${Reducer.CARDS_SET_FILTER} correctly`, () => {
      const {
        CARDS_SET_FILTER,
        defaultState,
      } = Reducer;
      const action = {
        type: CARDS_SET_FILTER,
        payload: {
          ...defaultState(),
          filters: { test: 'test' },
        },
      };
      const result = Reducer.default(undefined, action);
      expect(result).toEqual({
        ...defaultState(),
        filters: { test: 'test' },
      });
    });

    it(`handles ${Reducer.CARDS_SET_FILTER} with null payload correctly`, () => {
      const {
        CARDS_SET_FILTER,
        defaultState,
      } = Reducer;
      const action = {
        type: CARDS_SET_FILTER,
      };
      const result = Reducer.default(undefined, action);
      expect(result).toEqual({ ...defaultState() });
    });
  });
});
