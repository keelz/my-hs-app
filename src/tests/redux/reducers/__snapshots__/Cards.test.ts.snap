// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards Reducer snapshots composes default state correctly 1`] = `
Object {
  "activeClassName": "DRUID",
  "data": Array [],
  "filters": Object {},
  "pagination": Object {
    "currentPage": 0,
    "itemsPerPage": 0,
    "pages": 0,
    "total": 0,
  },
}
`;

exports[`Cards Reducer snapshots composes state correctly 1`] = `
Object {
  "activeClassName": "DRUID",
  "data": Array [],
  "filters": Object {},
  "pagination": Object {
    "currentPage": 0,
    "itemsPerPage": 0,
    "pages": 0,
    "total": 0,
  },
}
`;
